Fix for Rails 8.1 counter cache change

romanbsd opened the pull request:
This is my take on the counter cache issue in Rails 8.1

Feedback appreciated!

romanbsd committed:
Fix for Rails 8.1 counter cache change

romanbsd committed:
fix build, remove EOL Rails 7.0, add Rails 8.1

romanbsd commented:
Removed Rails 7.0, added 8.0 to matrix.

romanbsd commented:
oh, right, it cannot build because of node-sass...

timdiggins commented:
> oh, right, it cannot build because of node-sass...

feel free to rebase this one on your other PR #1017 (which should get it green) -- I will review this one once that is merged...

timdiggins commented:
I'm not 100% sure we should be working round this only in the TopicForm - it leaves it impossible to create a topic in the console. (I think you need to copy in this spec:  https://github.com/thredded/thredded/pull/993/changes#diff-9cf83f58efcb5c0700e5dd5047f98419dd8f188bd73cd7038605f56580886a92 and then fix it one way or another. 

I'm thinking maybe the best route is to drop the unnecessary extra foreign key (as mentioned in the linked spec) -- we should be able to just allow post → messageboard via a `through: :postable` (it's just the code is quite confusing because of some over-DRYing of post/private post)

----

If we go ahead with this one, it needs a rebase now on main, but also the rails 8.1 gemfile isn't quite right. Here are my tweaks.

```
gem 'rails', '~> 8.1.1'
gem 'rails-i18n', '~> 8.1.0'

# TODO: create a new dummy using import-maps ?
gem 'webpacker', '~> 5.0'

# https://github.com/rails/rails/blob/v8.1.1/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb#L14
gem 'sqlite3', '>= 2.1'

# https://github.com/rails/rails/blob/v8.1.1/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L3
gem 'pg', '~> 1.1'

# https://github.com/rails/rails/blob/v8.1.1/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb#L6
gem 'mysql2', '~> 0.5'
```

romanbsd committed:
Fix promote_errors

In Rails 6.1+, errors.each yields ActiveModel::Error objects, not [attribute, message] pairs

romanbsd commented:
I rebased meanwhile and fixed another issue
